crates/server/src/verifywalletsignatures/starknet.rs

```diff
@@ -0,0 +1,320 @@
+        .call(&function_call, BlockId::Tag(BlockTag::Latest))
+        .await;
+
+    match result {
+        // If the signature is valid, verify the hash
+        Ok(_) => {
+            let verify = verify_signature_hash(message_hash, wallet_address, message, chain_id);
+            if verify.is_ok() {
+                return Ok(true);
+            }
+            eyre::bail!("Invalid message hash");
+        }
+        Err(err) => {
+            eyre::bail!("Error verifying signature: {:?}", err);
+        }
+    }
+}
+
+// Function to verify a MetaMask Snap wallet signature off chain
+pub fn verify_metamask_signature(
+    message_hash: &str,
+    signature: Vec<String>,
+    signing_key: &str,
+    message: &str,
+    wallet_address: &str,
+    chain_id: &str,
+) -> eyre::Result<bool> {
+    // Convert inputs to Felt types
+    let signing_key = Felt::from_str(&signing_key)?;
+    let message_hash = Felt::from_str(&message_hash)?;
+    let wallet_address = Felt::from_str(&wallet_address)?;
+
+    // Verify the signature using the StarkNet crypto library
+    let result = verify(
+        &signing_key,
+        &message_hash,
+        &Felt::from_str(&signature[0])?,
+        &Felt::from_str(&signature[1])?,
+    );
+    match result {
+        // If the signature is valid, verify the hash
+        Ok(true) => {
+            let verify = verify_signature_hash(message_hash, wallet_address, message, chain_id);
+            if verify.is_ok() {
+                return Ok(true);
+            }
+            eyre::bail!("Invalid message hash");
+        }
+        Ok(false) => {
+            eyre::bail!("Invalid signature");
+        }
+        Err(err) => {
+            eyre::bail!("Error verifying signature: {:?}", err);
+        }
+    }
+}
+
+// Function to verify the integrity of a message hash by hashing the message and comparing it to the provided hash
+fn verify_signature_hash(
+    message_hash: Felt,
+    wallet_address: Felt,
+    message: &str,
+    chain_id: &str,
+) -> eyre::Result<()> {
+    let types = Types {
+        stark_net_domain: vec![
+            FieldType {
+                name: "name".to_string(),
+                field_type: "shortstring".to_string(),
+            },
+            FieldType {
+                name: "chainId".to_string(),
+                field_type: "felt".to_string(),
+            },
+            FieldType {
+                name: "version".to_string(),
+                field_type: "shortstring".to_string(),
+            },
+            FieldType {
+                name: "revision".to_string(),
+                field_type: "shortstring".to_string(),
+            },
+        ],
+        challenge: vec![
+            FieldType {
+                name: "nodeSignature".to_string(),
+                field_type: "string".to_string(),
+            },
+            FieldType {
+                name: "publicKey".to_string(),
+                field_type: "string".to_string(),
+            },
+        ],
```

Comment:
will `start_net_domain` always contain 4 items? and `challenge` 2? if so, you can represent that in the type..._